src/main/java/in/org/projecteka/hdaf/link/link/Link.java

```diff
@@ -3,53 +3,97 @@
+                        .flatMap(url -> hipClient.linkPatientCareContext(linkReferenceRequest, url)
+                                .flatMap(linkReferenceResponse -> {
+                                    linkReferenceResponse.setTransactionId(patientLinkReferenceRequest.getTransactionId());
+                                    return linkRepository.insertToLinkReference(
+                                            linkReferenceResponse, hipId)
+                                            .then(Mono.just(
+                                                    PatientLinkReferenceResponse.builder()
+                                                            .transactionId(
+                                                                    linkReferenceResponse.getTransactionId())
+                                                            .link(linkReferenceResponse.getLink()).build()));
+                                })
+                        )).switchIfEmpty(Mono.error(ClientError.unableToConnectToProvider()));
     }
 
-    public Mono<PatientLinkResponse> verifyToken(String linkRefNumber, PatientLinkRequest patientLinkRequest) {
-        //from linkRefNumber get TransactionId
-        //from transactionID get providerID
-        String providerId = "10000005";
-        //Check otp for expiry
-        return providerUrl(providerId)
-                .flatMap(url -> hipClient.validateToken(linkRefNumber, patientLinkRequest, url))
-                .switchIfEmpty(Mono.error(ClientError.unableToConnectToProvider()));
+    public Mono<PatientLinkResponse> verifyToken(String linkRefNumber, PatientLinkRequest patientLinkRequest, String patientId) {
+        return isOTPExpired(linkRefNumber).flatMap(expiry -> {
+            if(!expiry){
+                return linkRepository.getTransactionIdFromLinkReference(linkRefNumber)
+                        .flatMap(linkRepository::getHIPIdFromDiscovery)
+                        .flatMap(hipId -> providerUrl(hipId)
+                                .flatMap(url -> hipClient.validateToken(linkRefNumber, patientLinkRequest, url)
+                                        .flatMap(patientLinkResponse -> linkRepository.insertToLink(
+                                                hipId,
+                                                patientId,
+                                                linkRefNumber,
+                                                patientLinkResponse.getPatient()
+                                        ).then(Mono.just(
+                                                PatientLinkResponse.builder().patient(patientLinkResponse.getPatient()).build()))))
+                                .switchIfEmpty(Mono.error(ClientError.unableToConnectToProvider())));
+                }
+                return Mono.error(ClientError.otpExpired());
+        });
     }
 
     private Mono<String> providerUrl(String providerId) {
         return clientRegistryClient.providerWith(providerId)
```

Comment:
This code is duplicated in both discovery and link. we should think about extracting it somewhere.